File: md5dir.rb
#!/usr/bin/env ruby
$:.unshift File.join(ENV['SCRIPTS'], 'lib') unless ENV['SCRIPTS'] == nil
$:.unshift File.join(File.dirname(__FILE__), 'lib')

require 'dirglob'
require 'md5sum'

class MD5Dir
  def md5puts(name)
    f = File.open(name,"rb")
    t = File.stat(name)
    s = MD5Sum.new
    puts "MD5:#{s.md5sum(f)} #{t.mtime.strftime("%Y/%m/%d-%X")} #{t.size} #{name}"
    f.close
  end

  def dir(argv0, sub_directory)
    DirGlob.dir(argv0, sub_directory).each do |f|
      md5puts(f)
    end
  end
end

if __FILE__ ==$0
  require 'optparse'
  sub_directory = false

  parser = OptionParser.new do |parser|
    parser.banner = "#{File.basename($0,".*")} by id774 <idnanashi@gmail.com>
    Usage: #{File.basename($0,".*")} [path]"
    parser.separator "options:"
    parser.on('-d', '--subdirectory', "include sub directory"){ sub_directory = true }
    parser.on('-h', '--help', "show this message"){
      puts parser
      exit
    }
  end

  begin
    parser.parse!
  rescue OptionParser::ParseError => err
    $stderr.puts err.message
    $stderr.puts parser.help
    exit 1
  end

  m = MD5Dir.new
  if ARGV.size >= 1
    m.dir(ARGV[0], sub_directory)
  else
    m.dir(".", sub_directory)
  end
end

File: md5dir.py
#!/usr/bin/env python

import sys
import os
p = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'lib')
if not p in sys.path:
    sys.path.append(p)
try:
    p = os.path.join(os.environ['SCRIPTS'], 'lib')
    if not p in sys.path:
        sys.path.append(p)
except KeyError:
    pass

def md5_dir(subdirectory, args):
    import md5sum
    import os
    from stat import S_ISDIR, ST_MODE
    m = md5sum.Md5sum()
    if subdirectory or S_ISDIR(os.stat(args[0])[ST_MODE]):
        for root, dirs, files in os.walk(args[0]):
            for file in files:
                print((m.get_md5(root + os.sep + file),
                       root + os.sep + file))
    else:
        for f in args:
            print((m.get_md5(f), f))

def main():
    from optparse import OptionParser
    usage = "usage: %prog [options] file"
    parser = OptionParser(usage)
    parser.add_option("-d", "--subdirectory", help="include sub directory",
                      action="store_true", dest="subdirectory")
    (options, args) = parser.parse_args()
    if len(args) < 1:
        parser.print_help()
    else:
        md5_dir(options.subdirectory, args)

if __name__ == '__main__':
    main()
